### ISSUE TITLE
[CEP] In product custom reporting PoC

### ISSUE THREAD

**Issue by czue** [dimagi/commcare-hq#28078]
**Abstract**

*Update July 16, 2020: this CEP [has changed](https://github.com/dimagi/commcare-hq/issues/28078#issuecomment-657581111) since originally authored. Please review the comment history before commenting. Also this CEP now depends on this related CEP: https://github.com/dimagi/commcare-hq/issues/28140*

This proposal describes an architecture for building complex data visualizations inside CommCare HQ. It is being drafted to support our Covid Analytics work, and the idea is that it would be tested/used to create an initial set of reports for our domestic COVID projects - likely at the county level.

This CEP has been ported from [this Google Doc](https://docs.google.com/document/d/16agZrznml_K-T8DYp8Da1PBFWhdJ0AJKH7w8ay3FQxk/edit#), but comments can be made **here in the github issue** - as the document contains non-essential information and references to other work in flight.

There is a reference implementation of just the front-end side of this available for discussion here: https://github.com/dimagi/commcare-hq/pull/28086/

**Motivation**

The motivation is to create a mechanism for delivering complex, in-product visualizations that break out of the normal CommCare reporting paradigms, which are not only constraining but also difficult to work in.

**Specification**

The basic idea is to create a client/server architecture inside HQ as follows:
![image](https://user-images.githubusercontent.com/66555/87047893-5e397f80-c1fb-11ea-87b7-c3218713165c.png)

*The Front End*

The front end should be a standalone client-side web page built specifically for the needs of the project. **It should contain no Python / Django code and be exclusively JavaScript and HTML.** It should also be fully custom.

Some reasons not to try to make the front end generic or modular and making it a fully custom app include:

- A custom app will rapidly accelerate the rate of development and change possible.
- Custom apps allow different development teams to leverage front-end technologies they are most comfortable with.
- We believe that the user-facing needs of complex reports will inevitably require breaking any attempt at generalization we do create.
- We are explicitly not attempting to build a BI tool in CommCare due to the complexity and maintenance burden that would entail.
- Even if we did want to build a generic solution - which we don’t - we would still want to learn from a few one-offs before attempting to do that.

In practice, the front end will be a Django page served by CommCare HQ, but the page should basically be a blank canvas for front-end JavaScript to take over. This allows us to leverage CommCare HQ’s permissions, session management, and authentication while providing the flexibility we desire.

In an ideal world the front end should be written in a modern JavaScript framework like React, or Vue.js, though in practice we may choose Knockout.js - or no framework at all - to leverage the Dimagi team’s current knowledge and familiarities.

Eventually these reports should be written in the [plug in framework](https://github.com/dimagi/commcare-hq/pull/28050) so as to avoid polluting the HQ code base and the environments where they are not needed.

*The Back End*

To start **we will try to build the back end 100% off of UCR based on the existing [Report Data API](https://confluence.dimagi.com/display/commcarepublic/Download+Report+Data)**. In the longer term, we may wish to expand the back end to include other APIs and data source types.

**Impact on users**

This will have no impact on users except those who are able to use the custom reports created within the framework.

**Impact on hosting**

This may impact code footprint, depending on the degree to which we are able to separate the custom front ends (or not) using something like a plug in framework. However the code in question should be 100% isolated and not relevant to any HQ deployments not using the reports.

**Backwards compatibility**

This CEP is 100% backwards compatible.

**Release Timeline**

We are hoping to implement a PoC of this by the end of July 2020.

**Open questions and issues**

- Is there an easy way to isolate the front-end code such that it doesn't reside in the main commcare-hq repository?
- How can we manage the dependencies between the front-end and the UCRs? Or should we require the UCRs be static to start?
- How can we ensure the security of any third-party JavaScript does not compromise the security / integrity of HQ?
- How can we ensure we do not hamper the ability for us to maintain our own front-end code without having the burden of maintaining and testing these external libraries long term?

**Comment by snopoke** [comment]
> How can we manage the dependencies between the front-end and the UCRs? Or should we require the UCRs be static to start?

If the UI makes use of the Report Data API then it could be possible to create new reports off the same datasource instead of modifying the old report. Then once the report is ready the UI can be updated to use the new report. Not a great process but we don't have another way to version the reports or data sources.

> Is there an easy way to isolate the front-end code such that it doesn't reside in the main commcare-hq repository?

Not at this point. The current work is looking to allow additional git repos to reside alongside commcare. For this feature specifically that seems like overkill. If the code was packaged correctly could we piggy back on any of the existing dependency management tools we have?

Another (possibely sketchy option) would be to zipp the UI code and uploaded to the CommCare. It could then be updated via the web without any code changes. Alternately you could host it elsewhere and just tell CommCare where to get it.

Thinking about this more I'm concerned about security issues of running potentially 3rd party javascript code as part of CommCare. If we are the ones developing it and have full control over the deployment of the code then it's not an issue but if it become a certainly if we did either of the things I mentioned above I'd be concerned about security.

**Comment by czue** [comment]
not really sure who specifically should review this, so ping @dimagi/dimagi-dev

**Comment by millerdev** [comment]
> The Front End ... be exclusively JavaScript and HTML.
>
> Custom apps allow different development teams to leverage front-end technologies they are most comfortable with.

How will JavaScript dependencies of these front-end components be managed? We already struggle to keep our front-end libraries up to date (just ask @biyeun). I would not like to have a bunch of forgotten third-party-developed code that depends on outdated JS libs holding us back when we attempt to bring ourselves up to date in three years from now.

@snopoke said
> Another (possibely sketchy option) would be to zipp (sic) the UI code and uploaded to the CommCare.

I think this is the definition of a code injection vulnerability. How will the code be audited?

I would like to propose a less tightly integrated approach where all of the front end code resides in external projects, is deployed separately, and consumes HQ data through APIs. We may want to build a separate "hosting project" for these so each one does not need to do all the work of building its own deployment framework, etc., but I think it should not be part of the current HQ code base. I think the biggest impediment here is authentication/authorization for API access (front-end-only code cannot easily authenticate without exposing keys to the client). I'm happy to brainstorm about how this can be done if needed.

**Comment by orangejenny** [comment]
I like the thought of this work interacting with HQ via APIs. Sharing front end dependencies - js, styling, or HTML - would be a nightmare.

**Comment by czue** [comment]
> How will JavaScript dependencies of these front-end components be managed? 

I was imagining a separate build / toolchain where the output file is all that's dropped into HQ. I think it will be easier to explain in code once I have a PoC but basically each project can have it's own build pipeline with whatever tooling it wants, generates the bundles it needs, and HQ just drops an (unmanaged by HQ build tooling) static file on a page. It should not have dependency on how HQ wants to handle JavaScript at all and the increased page weight would only impact the reports it is used on. In theory the entire thing could be a separate repo, though that then gets into the code / security questions raised by @snopoke 

> I would like to propose a less tightly integrated approach where all of the front end code resides in external projects, is deployed separately, and consumes HQ data through APIs

This sounds interesting, though I'm unclear exactly how it would work. One of the key requirements of the work is to provide a seamless integration with HQ - which means it needs to "feel" embedded in the site. If there is a one-time auth I could imagine that being ok, but anything beyond that I think would be untenable (including redirecting people to a site hosted elsewhere - though iframes may work).

@millerdev @orangejenny given the clarification on JS dependencies do you still have concerns that would prevent a PoC in the stated approach?

**Comment by millerdev** [comment]
> it needs to "feel" embedded in the site

I don't see how this will work without sharing our base templates at least, which includes many of our JS libraries and CSS. Do you have an idea of how to get around that?

On the security side, since it is served from commcarehq.org it will have the same level of access to cookies/session keys as any other page, and therefore if one of these pages used a library that had, for example, an [XSS vulnerability](https://owasp.org/www-community/attacks/xss/) it could allow an untrusted party to steal session keys from unsuspecting users of the site.

**Comment by czue** [comment]
> I don't see how this will work without sharing our base templates at least, which includes many of our JS libraries and CSS. Do you have an idea of how to get around that?

I don't, though can you articulate why that would be a problem?

> On the security side, since it is served from commcarehq.org it will have the same level of access to cookies/session keys as any other page, and therefore if one of these pages used a library that had, for example, an XSS vulnerability it could allow an untrusted party to steal session keys from unsuspecting users of the site.

Agreed. I think in the short term the proposal would be to push these through standard code review policies so we don't have to figure out exactly how to solve this for now.

**Comment by millerdev** [comment]
> I don't, though can you articulate why that would be a problem?

We can't make backward-incompatible changes to our templates without possibly breaking these pages, as just one example.

**Comment by czue** [comment]
I'm going to schedule a meeting to discuss this approach with the broader team. Please ping me if you'd like to be added (will invite anyone who comments on this issue).

i also PR'd [a sketch of what this might look like](https://github.com/dimagi/commcare-hq/pull/28086/) to help make sure we are starting from the same going-in position in terms of what an implementation might look like.

I also added two more points to the open issues section that I believe encapsulate the concerns raised so far. @millerdev @orangejenny let me know if I missed anything.

**Comment by biyeun** [comment]
@czue please add me to this meeting too.

**Comment by czue** [comment]
Added you - it's a little early but the only time that worked for the four of us already

**Comment by orangejenny** [comment]
@czue Thank you for setting up a conversation. In advance of that, here are a few things I'm going to want to talk about:

- I'm surprised you're considering sharing the base template; I thought about asking earlier but then took for granted that of course we wouldn't share it. The base template makes a number of assumptions about standard javascript practices that are expected to be followed across HQ, the biggest two being which 1) javascript libraries are included globally and 2) how js dependencies are managed.
- JS dependencies on HQ are currently half-migrated to RequireJS, which is a fragile state that I don't want to make worse. At this point RequireJS itself isn't an especially modern technology; I'm anticipating that we should not only finish the migration but also, possibly at the same time, shift to something like webpack, which could also have implications for how we handle other static assets. What kinds of expectations would there be for this new code's static asset management, and who would be responsible for keeping it in line with HQ's standards?
- Similarly, Biyeun has been interested for a while in shifting from LESS to SaaS, that's another major change that would break downstream code.
- Historically, I don't think our standard code review process has worked well for large front end "prototypes." There has been too much code to review effectively and there's been strong pressure to get it merged fast. I'm thinking of the original ICDS dashboard and also of the ICDS mobile dashboard. I put a lot of time into reviewing the latter but still don't think it was up to normal code review standards.
- Besides the ICDS dashboard, web apps is another example of a large piece of front end code that is partially but not completely independent of HQ, which has made it problematic from a maintenance, modernization, and security perspective.

**Comment by czue** [comment]
@orangejenny can you confirm you either have reviewed or plan to review the companion PR to this in addition to all my comments? I think it obsoletes several of these questions regarding the coupling with HQ and the dependencies (though some points are definitely still valid).

I'm not married to the idea of sharing a base template if we think that will be problematic. I don't expect it to be, but that may be naive.

**Comment by orangejenny** [comment]
@czue Sure, I can look at the PR before Monday. I skimmed it and the usage of script tags was what caught my eye.

**Comment by czue** [comment]
Sorry last comment in case helpful. I think one question for you and Daniel
going in is whether you are rejecting the premise that complex custom
reports should exist in HQ full stop, or whether it's this particular
implementation that's problematic. If it's the latter I'm very open to
alternative implementations - I just haven't wrapped my head around how
they might work on a practical timeline.

And thanks for the questions - I will think about them and try to get
something written down prior to Monday.

**Comment by orangejenny** [comment]
One alternative implementation, which I haven't thought through in detail, would be using an iframe for the custom report.

**Comment by orangejenny** [comment]
Just re-read the email thread and realized you mentioned iframes above, missed it on my first read.

**Comment by czue** [comment]
FYI I tried to draft a framing for Monday's discussion. There is an [editable version here](https://docs.google.com/document/d/16agZrznml_K-T8DYp8Da1PBFWhdJ0AJKH7w8ay3FQxk/edit#heading=h.pdx1nv5lwal6).

**Motivation and Requirements**

- It is expected we will need to build a reporting stack to meet the data needs for our domestic covid projects
- Reports need to be more visually complex and interactive than what can be achieved in CommCare’s current paradigms.
- We seek to provide a user experience that is as seamless as possible for our existing CommCare users. Requiring managing multiple accounts is highly discouraged.
- We would like to maximize the speed of development of these reports, and ideally would have something to show in the next few weeks.
- We would like these reports to be authorable and maintainable by non-Dimagi developers.
- We seek to minimize the amount of overhead introduced to maintain these reports - especially to Dimagi’s core team and core codebase. Ideally there would be zero increased maintenance burden.
- We seek to minimize the possible disruption to the core product caused by these reports, including impact on performance and security. Ideally there would be zero increased performance or security risk from a product perspective.
- Conversely, it would be preferable if the reports were robust to underlying changes in the core product codebase.

*Do we agree on these? Is the list complete?*

**Possible paths forward**

- Try “Standard” custom reports
- Try the proposed architecture
- Try a third party site (possibly embedded in an iframe)
- Reject the goals of the project outright and make the stakeholders deal with the consequences

*Are there others that should be on the table? If so what are they?*

**Evaluation / Discussion**

How do the proposed architectures meet the goals stated above? What’s realistic and what’s not and where can we make compromises?

Can we agree on the right candidate proposed architecture to pursue further so that we can move onto mitigation strategies to address the issues (since there are likely to be issues with all choices)?

Mitigation strategies and discussions based on architecture decisions. This will be dependent on getting to high-level agreement on a path forward (e.g. it’s not worth debating the merits of whether to extend the base template if we decide to use in-product reports or a third party site). 

Depending on what we decide, this would be the time to get into details on a lot of the issues raised in the CEP.

@orangejenny @biyeun @millerdev @snopoke

**Comment by czue** [comment]
@orangejenny here's my quick reaction to your comments just so you have them, but I'm hoping to start one higher level up on Monday's discussion.

> I'm surprised you're considering sharing the base template; I thought about asking earlier but then took for granted that of course we wouldn't share it. The base template makes a number of assumptions about standard javascript practices that are expected to be followed across HQ, the biggest two being which 1) javascript libraries are included globally and 2) how js dependencies are managed.

I'm unclear whether or not this will be a problem in practice. My impression is that a library like React operates very specifically on its own managed DOM components in a way that would not be harmed by the existence of other stuff on the page. Either way, I think there are pretty easy mitigation strategies on this one, e.g. don't use the base template, fork it and remove the JS, etc.

> JS dependencies on HQ are currently half-migrated to RequireJS, which is a fragile state that I don't want to make worse. At this point RequireJS itself isn't an especially modern technology; I'm anticipating that we should not only finish the migration but also, possibly at the same time, shift to something like webpack, which could also have implications for how we handle other static assets. What kinds of expectations would there be for this new code's static asset management, and who would be responsible for keeping it in line with HQ's standards?

In practice I am proposing that this code use webpack. But in terms of your question, I think that's an open discussion. Two possible answers: either HQ can declare what these libraries should use, or HQ can not concern itself at all. I favor the latter, but understand why it can cause uneasiness.

> Similarly, Biyeun has been interested for a while in shifting from LESS to SaaS, that's another major change that would break downstream code.

Again, I'm not sure this is true if we use properly scoped styles (readily available in most modern JS these days). But even if it is, I think there are pretty easy mitigation strategies similar to the first point.

> Historically, I don't think our standard code review process has worked well for large front end "prototypes." There has been too much code to review effectively and there's been strong pressure to get it merged fast. I'm thinking of the original ICDS dashboard and also of the ICDS mobile dashboard. I put a lot of time into reviewing the latter but still don't think it was up to normal code review standards.

Agreed. One question I have is whether we can set this up in a way that this is not actually a problem (i.e. figure out a way so that the HQ team doesn't really have to care about the code quality of these). I think this is a big question and I don't have a clear answer today. Do you know how you'd propose doing the ICDS dashboard front-end today if we were starting over?

> Besides the ICDS dashboard, web apps is another example of a large piece of front end code that is partially but not completely independent of HQ, which has made it problematic from a maintenance, modernization, and security perspective.

Noted.

**Comment by czue** [comment]
Update from today's discussion:

We will attempt to move forward with a more decoupled approach that involves:

- A third-party static site on a different domain that accesses CommCare HQ via APIs only
- Authentication managed by CommCare HQ (some work might need to be done to do this properly - e.g. make HQ an oauth provider)
- API keys stored in a cookie (we may need to add some expiry and lifecycle management to support this with our required security SLAs)
- Exploration of embedding this site in an iframe in HQ for an improved user experience

If the above proves to be infeasible on the needed timeline we will revisit other approaches and try to come up with a solution that gets us in the direction of the above architecture and optimizes around the project and maintenance constraints as best we can.

**Comment by czue** [comment]
this is superseded by https://github.com/dimagi/commcare-hq/issues/28140
